### PR TITLE
feat: salted hashing

### PR DESCRIPTION
**Pull request by cmdoret**
## Proposed Changes

* Adds a `--secret` CLI arg to provide a file with encryption key or hashing salt.
* Reduce implementation surface for the `Pseudonymize` to a single method (`pseudo(bytes) -> String`).
* helper function to instantiate hasher based on algorithm and key, with support for default settings.

## Types of Changes

What types of changes does your code introduce? _Put an `x` in the boxes that
apply_

- [ ] A **bug fix** (non-breaking change which fixes an issue).
- [x] A new **feature** (non-breaking change which adds functionality).
- [x] A **breaking change** (fix or feature that would cause existing
      functionality to not work as expected).
- [ ] A **non-productive** update (documentation, tooling, etc. if none of the
      other choices apply).

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask. We're here to
help! This is simply a reminder of what we are going to look for before merging
your code._

- [x] I have read the
      [CONTRIBUTING](https://github.com/sdsc-ordes/tripsu/blob/master/CONTRIBUTING.md)
      guidelines.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added the necessary documentation (if appropriate).

### CODE REVIEW

**Comment by cmdoret** [comment]
Example use:
```sh
➜ tripsu pseudo -s secret.txt -c tests/data/config.yaml -x tests/data/type_map.nt tests/data/test.nt | tail -n3
<http://example.org/7c73d8167447e118ee76295f252e9ceeb48ae3cbfa0454572381796f0e7a8e2c> <http://schema.org/name> "f4446c44c9b6e221e94a40476322dab5123514c5669092a142057a3322329c09" .
<http://example.org/Bank> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://xmlns.com/foaf/0.1/Organization> .
<http://example.org/Bank> <http://schema.org/name> "Bank" .

➜ tripsu pseudo -c tests/data/config.yaml -x tests/data/type_map.nt tests/data/test.nt | tail -n3 
<http://example.org/e72357361fec5fc2f653642e6a1016040aa79097ab123468f3f7873907211d26> <http://schema.org/name> "f4aab5970bf4589ea821964476f10ddcef7b2a42a4796954f5cb32426bc5ec55" .
<http://example.org/Bank> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://xmlns.com/foaf/0.1/Organization> .
<http://example.org/Bank> <http://schema.org/name> "Bank" .
```

File: src/crypto.rs [hunk 1]
```diff
@@ -60,13 +57,58 @@ impl Pseudonymize for DefaultHasher {
...
+        let mut new_key = [0u8; 32];
+        if key.len() == 32 {
+            new_key.copy_from_slice(&key[..32]);
+        } else {
+            panic!("Key must be 32 bytes long");
+        }
+
+        return Self { key: new_key };
+    }
+}
```
Comment (martinfontanet) [inline src/crypto.rs:108]: - Change input parameter to `Optional<Vec<u8>>`
- Move `key.unwrap_or(...)` here
- Check minimal key length depending on the algorithm
- Hash key to have the correct key length

Comment (cmdoret) [reply src/crypto.rs]: Good points: all are now implemented

File: src/crypto.rs [hunk 2]
```diff
@@ -78,9 +80,7 @@ impl Default for Algorithm {
     }
 }
 
-pub fn get_pseudonymizer(algo: Option<Algorithm>, key: Option<Vec<u8>>) -> impl Pseudonymize {
-    let secret = key.unwrap_or(generate_key(32));
-
+pub fn get_pseudonymizer(algo: Option<Algorithm>, secret: Option<Vec<u8>>) -> impl Pseudonymize {
```
Comment (martinfontanet) [inline src/crypto.rs:83]: Maybe this is not relevant for this PR, but shouldn't we rename this into `new_pseudonymizer` or use a method of `Pseudonymizer` instead (`Pseudonymizer::new`)?
The function creates a new pseudonymizer, it does not really "get" it.

File: src/crypto.rs [hunk 3]
```diff
@@ -2,6 +2,8 @@ use super::model::Entity;
 use crate::{model::TripleMask, rdf_types::*};
 use rand::Rng;
 
+const MIN_SECRET_SIZE: usize = 32;
```
Comment (martinfontanet) [inline src/crypto.rs:5]: I think this should be defined at the algorithm level, not globally. Some algorithms may require different min secret sizes. For example AES can use 128bits keys.

File: src/crypto.rs [hunk 4]
```diff
@@ -95,15 +95,22 @@ pub struct Blake3Hasher {
...
-        }
+    pub fn new(secret: Option<Vec<u8>>) -> Self {
+        secret.as_ref().inspect(|s| {
+            if s.len() < MIN_SECRET_SIZE {
+                panic!("Secret must be at least {MIN_SECRET_SIZE} bytes long");
+            }
+        });
+
+        // blake3 key must be exactly 32 bytes long
+        let mut key = [0u8; 32];
```
Comment (martinfontanet) [inline src/crypto.rs:106]: I don't know if this is possible in Rust, but could we have something like `let mut key = [0u8; MIN_SECRET_SIZE]` instead? Just to make sure we stay consistent with the key size. The same applies for the lines bellow (lines 108, 109, 111)

**Comment by cmdoret** [comment]
Took your points into account and added tests + docstrings!

**Comment by cmdoret** [comment]
Just realized something: if we plan to support symmetric encryption, wouldn't `--secret` be the public key? That sounds a bit unintuitive

**Comment by martinfontanet** [comment]
> Just realized something: if we plan to support asymmetric encryption, wouldn't `--secret` be the public key? That sounds a bit unintuitive

For this, I think we should just change the behavior of tripsu  when asymmetric encryption is used and require `--public-key` or something similar instead of `--secret`. However, I suggest that we do it  **only if** one day we have a good reason to use asymmetric encryption.

It does not really make sense to me to support it for the following reasons:
- Speed: asymmetric encryption requires more computational power and is much slower for data stream encryption.
- Purpose: the goal of tripsu is to remove some information before sending the data, not to securely transfer it. The benefit of using asymmetric encryption is, in general, that you can send sensitive data in a non-trusted channel. For tripsu, we want to make sure that only the sender can decrypt the data, that's why we use either hash functions or symmetric encryption without sharing the key. If the data sender is the only one that encrypts and decrypts the data, there is no benefits of using asymmetric encryption.

**tldr;** I wouldn't worry about this for now :grin:

File: src/crypto.rs [hunk 5]
```diff
@@ -23,50 +36,132 @@ pub trait Pseudonymize {
...
+    fn test_generate_key() {
+        let key = generate_key(42);
+        assert_eq!(key.len(), 42);
+    }
+
+    #[test]
+    fn test_pseudo_named_node() {
+        let hasher = Blake3Hasher::new(None);
+        let named_node = NamedNode {
+            iri: "http://example.com/abc".to_string(),
```
Comment (martinfontanet) [inline src/crypto.rs:147]: `abc` is valid hex, so if we `pseudo_named_node` returns `named_node` without modifying it the test still passes :stuck_out_tongue: 
I suggest we change it to some non-hex string.
```suggestion
            iri: "http://example.com/tripsu".to_string(),
```

Comment (cmdoret) [reply src/crypto.rs:147]: nice catch! :sunglasses:

**Review by martinfontanet — APPROVED** [review]
